### PR TITLE
CLN: reorganize index.py, test_index.py

### PR DESCRIPTION
**Pull request by wesm**
Split apart these two very large modules and created a new `pandas.indexes` subpackage. It would be nice to move all the index class code from `pandas.tseries` there in a followup patch.

### CODE REVIEW

**Comment by wesm** [comment]
Once the build passes, best not to let this patch linger (to avoid rebase conflicts). @jreback I don't think I broke anything but let me know if any concerns that I can look into

**Comment by jorisvandenbossche** [comment]
The python 2 tests are still failing, I think you missed a `from pandas.compat import range` import in the `range.py` file

**Comment by wesm** [comment]
Thanks, think I got it now. Will check the build in an hour or two here

**Comment by wesm** [comment]
@jorisvandenbossche @jreback merge when ready. thank you

File: pandas/indexes/api.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
+from .base import (Index, _new_Index,  # noqa
+                   _ensure_index, _get_na_value,
+                   InvalidIndexError)
+from .category import CategoricalIndex  # noqa
+from .multi import MultiIndex  # noqa
+from .numeric import NumericIndex, Float64Index, Int64Index  # noqa
```
Comment (jreback) [inline pandas/indexes/api.py]: the relative imports are not consisten with the rest of the import methodology. Ok, but should define when relative imports are ok.

File: pandas/indexes/api.py [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+import pandas.core.common as com
+import pandas.lib as lib
+
+
+__all__ = ['Index', 'MultiIndex', 'NumericIndex', 'Float64Index', 'Int64Index',
+           'CategoricalIndex', 'RangeIndex',
+           'InvalidIndexError',
+           '_new_Index',
+           '_ensure_index', '_get_na_value', '_get_combined_index',
+           '_get_distinct_indexes', '_union_indexes',
```
Comment (jreback) [inline pandas/indexes/api.py]: any reason you are expliciting adding the 'private' index methods to the import space?

File: pandas/tests/indexes/common.py [hunk 3]
```diff
@@ -0,0 +1,678 @@
...
+from pandas import (Series, Index, Float64Index, Int64Index, RangeIndex,
+                    MultiIndex, CategoricalIndex, DatetimeIndex,
+                    TimedeltaIndex, PeriodIndex)
+from pandas.util.testing import assertRaisesRegexp
+
+import pandas.util.testing as tm
+
+import pandas as pd
+
+if PY3:
```
Comment (jreback) [inline pandas/tests/indexes/common.py]: why is this needed? shouldn't just be imported from ``pandas/compat`` directly?

File: pandas/tests/indexes/common.py [hunk 4]
```diff
@@ -0,0 +1,678 @@
...
+    unicode = lambda x: x
+
+
+class Base(object):
+    """ base class for index sub-class tests """
+    _holder = None
+    _compat_props = ['shape', 'ndim', 'size', 'itemsize', 'nbytes']
+
+    # @staticmethod
```
Comment (jreback) [inline pandas/tests/indexes/common.py]: was this originally here?

File: pandas/tests/indexes/test_base.py [hunk 5]
```diff
@@ -0,0 +1,1492 @@
...
+from pandas.util.testing import assert_almost_equal
+
+import pandas.core.config as cf
+
+from pandas.tseries.index import _to_m8
+
+import pandas as pd
+from pandas.lib import Timestamp
+
+if PY3:
```
Comment (jreback) [inline pandas/tests/indexes/test_base.py]: same here

File: pandas/indexes/api.py (hunk 1 above)
Comment (wesm) [inline pandas/indexes/api.py]: I'll make them absolute

File: pandas/indexes/api.py (hunk 2 above)
Comment (wesm) [inline pandas/indexes/api.py]: Other modules rely on these names existing in `pandas.core.index`

File: pandas/tests/indexes/common.py (hunk 3 above)
Comment (wesm) [inline pandas/tests/indexes/common.py]: i don't know, it was there before

File: pandas/tests/indexes/common.py (hunk 4 above)
Comment (wesm) [inline pandas/tests/indexes/common.py]: I started working on trimming unneeded computation from `setUp` (to speed up the tests -- e.g. I did this with the DataFrame tests and cut 30 seconds of test time) but didn't finish. Will do so some other time

Comment (jreback) [inline pandas/tests/indexes/common.py]: ahh ok, yeh I saw you added cached setup routines like fixtures in some other test modules

File: pandas/tests/indexes/common.py (hunk 3 above)
Comment (jreback) [inline pandas/tests/indexes/common.py]: hmm, ok, maybe add a TODO, thats just extraneous

File: pandas/indexes/api.py (hunk 2 above)
Comment (jreback) [inline pandas/indexes/api.py]: hmm, then they really should be importing directly (if you can gr8, otherwise pls add a TODO)

**Comment by wesm** [comment]
OK, cleaned up things per your comments.

File: pandas/indexes/api.py (hunk 2 above)
Comment (wesm) [inline pandas/indexes/api.py]: done

**Comment by jreback** [comment]
lgtm. ping when green.

**Comment by wesm** [comment]
@jreback we are good to go

**Comment by jreback** [comment]
boo ya! thanks!
